### ISSUE TITLE
Should the `thing-example.js` thing be setting the "reported" state instead of "desired"?

### ISSUE THREAD

**Issue by develpr** [aws/aws-iot-device-sdk-js#145]
I'm still wrapping my head around the intended use of `reported` vs `desired`, but based on my current understanding the `desired` state is basically "Thing, I would like you to change your state to this new state" - i.e. in the example of a remote IoT lightbulb and a phone app, the phone app would set the `desired` state of the thing to tell the light that it should go to that state.

The `reported` state is what the light would return back with, to let AWS IoT know that it was it's state. This could be done as a confirmation that the state had been moved to match the `desired` state, or it could be used to indicate that some external force caused the thing to go to that state (maybe somebody hit a remote control that changed the light to a new state).

*Again, I could be totally off on my understanding and would greatly appreciate being corrected if so!*, but if not, then the example `thing-example.js` seems to be a bit confusing because the thing part of the example sets the desired state. Again I'm probably thinking about htis all wrong, but it just seems to me like the "thing" should be setting the reported value?

**Comment by groteworld** [comment]
I think you are just missing a small part of a thing shadow's document which is the `delta`. If the `reported` doesn't match the `desired` a `delta` is create that states the desired. A thing can then handle changing to that missing `desired` state.

Example:

```js

{
  "desired": {
    "light": "green"
  },
  "reported": {
    "light": "blue"
  }
  "delta": {
    "light": "green"
  }
}
```

Here the shadow is stating there is a desired change, `delta`, and it will remain there until the `reported` state returns the same as the `desired`.

In the case of the example `thing-example.js` it is setting `desired` because it handles its "light" changing through the delta. since the `reported` (null) will never be `desired` (random RGB), the `delta` is always being updated and it's event can be listened for.

I hope this helps!!

**Comment by liuszeng** [comment]
Hi @develpr ,

Thank you very much for your interest in AWS IoT.

Thanks to @groteworld 's explanation. You can actually think about the following model:
1. You have a device with a light and a physical light controller on it.
2. You have a application that you would like to use to control that device light.
3. Both the device and your application runs AWS IoT Device SDKs, communicating with each other using MQTT.
4. The device light always updates the reported state, indicating this is exactly what happens on the device.
5. The application always updates the desired state, indicating that is exactly what you want the device to be.
6. The device light also listens to the delta messages, which are triggered whenever there is a difference between the desired and the reported state, that is to say, the device status is different from what you want it to be and needs to be changed.
7. The device light reacts to an incoming delta message by switching its status to the desired state you just requested through your application.

Basically, your device and you application are connected to each other through AWS IoT Device Shadow and its delta mechanism. You can control your device from your application by simply updating your shadow state.

For more details, you can take a look at the following documentations:
http://docs.aws.amazon.com/iot/latest/developerguide/thing-shadow-mqtt.html#update-delta-pub-sub-topic
http://docs.aws.amazon.com/iot/latest/developerguide/thing-shadow-data-flow.html

You can also take a look at a more use case related sample provided along with the Node.js SDK:
https://github.com/aws/aws-iot-device-sdk-js/tree/master/examples/temperature-control

Thanks,
Liusu

**Comment by id-mitchell** [comment]
Thank you @groteworld and @liuszeng for your response!

I THINK that your comment @liuszeng is inline with my understanding, but actually this is why I opened the issue. Your comment says:

> The *device* light always updates the reported state, indicating this is exactly what happens on the device.
> The *application* always updates the desired state, indicating that is exactly what you want the device to be.

but if you look in the `thing-example.js` example, when the device connects it runs this:

```function deviceConnect() {
      thingShadows.register(thingName, {
            ignoreDeltas: true
         },
         function(err, failedTopics) {
            if (isUndefined(err) && isUndefined(failedTopics)) {
               console.log('Device thing registered.');
               genericOperation('update', generateRandomState());
            }
         });
   }```

So this is what happens when the *device* (i.e. the "light" in our example) connects to AWS IoT, and if you look at `generateRandomState` you'll see:

```...snip...
state: {
            desired: rgbValues
         }
...snip...```

Basically the device is updating it's own desired state, instead of the reported state? Maybe I'm totally misunderstanding or mis-reading the code though?

What I would propose I guess is that the device update it's `reported` state, and the "app" set the `desired` state.

Again, apologies for wasting your time if I've just completely misread the code!

**Comment by develpr** [comment]
Thank you @groteworld and @liuszeng for your response!

I THINK that your comment @liuszeng is inline with my understanding, but actually this is why I opened the issue. Your comment says (emphasis mine!):

> The **device light always updates the reported state**, indicating this is exactly what happens on the device.
> The **application always updates the desired state**, indicating that is exactly what you want the device to be.

but if you look in the `thing-example.js` example, when the device connects it runs this:

```
function deviceConnect() {
      thingShadows.register(thingName, {
            ignoreDeltas: true
         },
         function(err, failedTopics) {
            if (isUndefined(err) && isUndefined(failedTopics)) {
               console.log('Device thing registered.');
               genericOperation('update', generateRandomState());
            }
         });
   }
```

So this is what happens when the *device* (i.e. the "light" in our example) connects to AWS IoT, and if you look at `generateRandomState` you'll see:

```
...snip...
state: {
            desired: rgbValues
         }
...snip...
```

Basically the device is updating it's own desired state, instead of the reported state? Maybe I'm totally misunderstanding or mis-reading the code though?

What I would propose I guess is that the device update it's `reported` state, and the "app" set the `desired` state.

Again, apologies for wasting your time if I've just completely misread the code! 

(reposted with correct github account! oops!)

**Comment by liuszeng** [comment]
Hi @develpr ,

Thank you very much for the information.

The `thing-example.js` is actually demonstrating the messaging flow for 2 clients updating the same cloud shadow documentation, where notifications can be sent via shadow delta messages. As per sample description:
>// Simulate the interaction of a mobile device and a remote thing via the
// AWS IoT service.  The remote thing will be a dimmable color lamp, where
// the individual RGB channels can be set to an intensity between 0 and 255.  
// One process will simulate each side, with testMode being used to distinguish 
// between the mobile app (1) and the remote thing (2).  The remote thing
// will update its state periodically using an 'update thing shadow' operation,
// and the mobile device will listen to delta events to receive the updated
// state information.

To make it easier, you can take a look at a similar sample in out Python Device SDK:
https://github.com/aws/aws-iot-device-sdk-python/tree/master/samples/basicShadow

For practical use cases, where you need a mobile application to control a device state remotely, we can still apply the model mentioned earlier. It is demonstrated in the Node.js temperature control example.

Hope the above helps.

Thanks,
Liusu

**Comment by fengsongAWS** [comment]
Thanks for your interest in AWS IoT Device SDK. We will close this issue for now and please feel free to open another thread if you have more questions.
